### PR TITLE
Fix/224 ordering strategy not as expected

### PR DESCRIPTION
**Pull request by ArdtK**
## Issue addressed
Solves #224 

## Code of conduct
- [x] I HAVE NOT added sensitive or compromised (test) data to the repository.
- [x] I HAVE NOT added vulnerabilities to the repository.
- [x] I HAVE discussed my solution with (other) members of the KOSWAT team.

## What has been done?
Instead of comparing pairwise which reinforcements to keep, create a product of all reinforcements and check which to remove.

### Checklist
- [x] Tests are either added or updated.
- [x] Branch is up to date with `master`.
- [x] Updated documentation if needed.

## Additional Notes (optional)
Add any additional notes or information that may be helpful.

### CODE REVIEW

File: koswat/strategies/order_strategy/order_strategy.py [hunk 1]
```diff
@@ -91,28 +91,25 @@ def split_reinforcements() -> tuple[
...
-        _sorted_pairs = pairwise(_sorted + _last)
-        _kept = (
-            list(
-                filter(lambda x: x is not None, map(check_reinforcement, _sorted_pairs))
-            )
-            + _last
+        _pairs = product(_sorted, _sorted + _last)
+        _to_remove = set(
+            filter(lambda x: x is not None, map(check_reinforcement, _pairs))
         )
```
Comment (Carsopre) [inline koswat/strategies/order_strategy/order_strategy.py:108]: computer wise I believe this to be better (and also simpler):

```python
        def check_reinforcement(
            pair: tuple[StrategyReinforcementInput, StrategyReinforcementInput],
        ) -> bool:
            return pair[0].base_costs > pair[1].base_costs and pair[0].ground_level_surface >= pair[1].ground_level_surface:

        for _pair in product(_sorted, _sorted + _last):
            if not check_reinforcement(_pair):
                continue
            _sorted.remove(_pair[0])

```

File: koswat/strategies/order_strategy/order_strategy.py [hunk 2]
```diff
@@ -91,28 +91,25 @@ def split_reinforcements() -> tuple[
...
+        for _reinforcement in _to_remove:
+            _sorted.remove(_reinforcement)
 
-        return [x.reinforcement_type for x in _kept]
+        return [x.reinforcement_type for x in _sorted + _last]
```
Comment (Carsopre) [inline koswat/strategies/order_strategy/order_strategy.py:112]: not sure why it's added now and not earlier (you are already doing a `product(_sorted, _sorted +_last)`.

**Review by Carsopre — CHANGES_REQUESTED** [review]
Algorithm can be simplified.

File: koswat/strategies/order_strategy/order_strategy.py (hunk 2 above)
Comment (ArdtK) [reply koswat/strategies/order_strategy/order_strategy.py:112]: The entries in `product` are only used to determine which need to be removed. `_last` is never removed and needs to be added here.

File: koswat/strategies/order_strategy/order_strategy.py (hunk 1 above)
Comment (ArdtK) [reply koswat/strategies/order_strategy/order_strategy.py:108]: The only downside of this solution is that a reinforcement could be flagged to be removed >1 time. `remove` will then give an exception, so in that case I need to add a check if it still exists.
With your solution the `__hash__` can be removed from `StrategyReinforcementInput`.
WDYT?

Comment (ArdtK) [reply koswat/strategies/order_strategy/order_strategy.py:108]: I adjusted the solution in line with your suggestion. My comment on removing >1 time doesn't apply, as `_sorted` is adjusted during the loop.

**Review by Carsopre — APPROVED** [review]

**Pull request merged**